rule/range-val-address.go

```diff
@@ -0,0 +1,113 @@
+package rule
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// RangeValAddress lints
+type RangeValAddress struct{}
+
+// Apply applies the rule to given file.
+func (r *RangeValAddress) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	var failures []lint.Failure
+
+	walker := rangeValAddress{
+		onFailure: func(failure lint.Failure) {
+			failures = append(failures, failure)
+		},
+	}
+
+	ast.Walk(walker, file.AST)
+
+	return failures
+}
+
+// Name returns the rule name.
+func (r *RangeValAddress) Name() string {
+	return "range-val-address"
+}
+
+type rangeValAddress struct {
+	onFailure func(lint.Failure)
+}
+
+func (w rangeValAddress) Visit(node ast.Node) ast.Visitor {
+	n, ok := node.(*ast.RangeStmt)
+	if !ok {
+		return w
+	}
+
+	value, ok := n.Value.(*ast.Ident)
+	if !ok {
+		return w
+	}
+
+	ast.Walk(rangeBodyVisitor{
+		valueID:   value.Obj,
+		onFailure: w.onFailure,
+	}, n.Body)
+
+	return w
+}
+
+type rangeBodyVisitor struct {
+	valueID   *ast.Object
+	onFailure func(lint.Failure)
+}
+
+func (bw rangeBodyVisitor) Visit(node ast.Node) ast.Visitor {
+	asgmt, ok := node.(*ast.AssignStmt)
+	if !ok {
+		return bw
+	}
+
+	for _, exp := range asgmt.Lhs {
+		e, ok := exp.(*ast.IndexExpr)
+		if !ok {
+			continue
+		}
+		if bw.isAccessingRangeValueAddress(e.Index) { // e.g. a[&value]...
+			bw.onFailure(bw.newFailure(e.Index))
+		}
+	}
+
+	for _, exp := range asgmt.Rhs {
+		switch e := exp.(type) {
+		case *ast.UnaryExpr: // e.g. ...&value
+			if bw.isAccessingRangeValueAddress(e) {
+				bw.onFailure(bw.newFailure(e))
+			}
+		case *ast.CallExpr:
+			if fun, ok := e.Fun.(*ast.Ident); ok && fun.Name == "append" { // e.g. ...append(arr, &value)
```

Comment:
limited to 'append' functions to reduce false positives. 
another option would be to reduce the confidence for non 'append' functions.

Reply:
I was wandering, why limiting to `append` and assignments? Is using the address of the range value variable not always suspicious?

Reply:
i summarized the problem with an example: https://play.golang.org/p/uAJAgz7apQv 

if the dev is using the address of the value after the loop (by adding it to a list, map or variable), there will be problems.
if the dev is using the address of the value right there in the loop (not storing it for later processing, e.g. adding it to a list, map, variable in the outer scope, everything is fine)